File: app/src/main/java/com/zionhuang/music/ui/adapters/SearchSuggestionAdapter.java
package com.zionhuang.music.ui.adapters;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.zionhuang.music.databinding.ItemSuggestionBinding;
import com.zionhuang.music.viewmodels.YouTubeSuggestionViewModel;

import java.util.Collections;
import java.util.List;

public class SearchSuggestionAdapter extends RecyclerView.Adapter<SearchSuggestionAdapter.ViewHolder> {
    private List<String> mDataSet = Collections.emptyList();
    private YouTubeSuggestionViewModel mViewModel;

    public SearchSuggestionAdapter(YouTubeSuggestionViewModel viewModel) {
        mViewModel = viewModel;
    }

    public void setDataSet(List<String> dataSet) {
        mDataSet = dataSet;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        ItemSuggestionBinding binding = ItemSuggestionBinding.inflate(inflater, parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(mDataSet.get(position));
    }

    @Override
    public int getItemCount() {
        return mDataSet.size();
    }

    public String getQueryByPosition(int position) {
        return mDataSet.get(position);
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        ItemSuggestionBinding binding;

        ViewHolder(ItemSuggestionBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(String query) {
            binding.setQuery(query);
            binding.executePendingBindings();
            binding.fillTextButton.setOnClickListener(v -> mViewModel.setQuery(binding.getQuery()));
        }
    }
}

File: app/src/main/java/com/zionhuang/music/ui/adapters/SearchSuggestionAdapter.kt
package com.zionhuang.music.ui.adapters

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.zionhuang.music.R
import com.zionhuang.music.extensions.inflateWithBinding
import com.zionhuang.music.ui.viewholders.SuggestionViewHolder

class SearchSuggestionAdapter(
        private val fillQuery: (query: String) -> Unit,
) : RecyclerView.Adapter<SuggestionViewHolder>() {
    private var dataSet = emptyList<String>()

    fun setDataSet(dataSet: List<String>) {
        this.dataSet = dataSet
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SuggestionViewHolder =
            SuggestionViewHolder(parent.inflateWithBinding(R.layout.item_suggestion), fillQuery)

    override fun onBindViewHolder(holder: SuggestionViewHolder, position: Int) =
            holder.bind(dataSet[position])

    override fun getItemCount(): Int = dataSet.size

    fun getQueryByPosition(position: Int): String = dataSet[position]
}